### PR TITLE
refactor: migrate associations to TypeScript

### PR DESCRIPTION
**Pull request by ephys**
This PR migrates the 'association' folder to TypeScript.

This PR brings makes very important changes to how associations work in Sequelize:

# List of changes

- Association names have been normalized: All associations are named, and their name always corresponds to their value in `Model.associations.x`.
- Associations are unique: It's not possible to replace an association once defined (breaking change).
- Associations are immutable: Once defined, they cannot be changed (breaking change).
- The way associations are determined when using `include` has changed (breaking change).
- Includes are now strict.

## Changes to how foreign keys are created

Instead of duplicating code across all associations, `BelongsTo` is the only association capable of adding Foreign Keys. All other associations delegate to BelongsTo:

- `HasMany` creates its inverse `BelongsTo` association
- `HasOne` creates its inverse `BelongsTo` association
- `BelongsToMany` creates its inverse `BelongsToMany` association, both of which create a `HasMany` association between Through and Source, which creates a `BelongsTo`

## Association names have been normalized

Sequelize 6 would consider these two associations to be different:

```typescript
Project.belongsTo(User);
Project.belongsTo(User, { as: 'User' });
```

This worked most of the time, but it was inconsistent internally. Both associations shared the same name and `Project.associations.User` be equal to the last association to have been defined. 
Even more inconsistently, `project.getUser` and other mixin methods would belong to the first association to have been defined.

After this PR, the above code will throw an error saying that the alias is already in use and to use another one.

## Associations are unique

The following code will now throw, because you can't replace an association that has already been defined anymore (your codebase would have been *very* fragile if you relied on this anyway, it's a very unsafe thing to do):

```javascript
Project.belongsTo(User);
Project.belongsTo(User);
```

## The way associations are determined when using `include` has changed

When specifying an `include` using a model+as, Sequelize 6 would search for an association that was defined with that model & as.

So these failed (on purpose):

```javascript
Project.belongsTo(User);

Project.findAll({
  include: { model: User, as: 'User' },
});

// ----

Project.belongsTo(User, { as: 'User' });

Project.findAll({
  include: { model: User },
});
```

these worked:

```javascript
Project.belongsTo(User);

Project.findAll({
  include: { model: User },
});

// ----

Project.belongsTo(User, { as: 'User' });

Project.findAll({
  include: { model: User, as: 'User' },
});
```

---

In Sequelize 7, the way it works is that if only one association has been defined with that model as the target, it will use that one, otherwise it will disambiguate using `as`.

**If you're using sequelize-typescript, that is already how it works as they override our system to implement this**

That means using `as` is exactly the same as using `association:` and I think we should update our documentation to promote the use of `association:` over the use of `model:`. i.e. instead of doing this:

```javascript
Project.findAll({
  include: { model: User, as: 'User' },
});
```

do this:

```javascript
Project.findAll({
  include: 'User',
});
```

## Associations are immutable

Before this PR, associations were a bit fuzzy: Defining an association on both sides of the association would attempt to merge and reconcile their options.

A good idea in theory, but these associations are applied as soon as the method is called. This means reconciliation may need to revert changes that were already done on other objects, like renaming a previously added foreign key on a model.

Determining which association was paired with which was also tricky.

With this PR, it's still possible to define association on both sides, but their options must be perfectly compatible. For instance this is not valid:

```typescript
User.belongsToMany(Countries, { foreignKey: 'user_id' });
Countries.belongsToMany(User);
```

but this is:

```typescript
User.belongsToMany(Countries, { foreignKey: 'user_id' });
Countries.belongsToMany(User, { otherKey: 'user_id' });
```

To make things easier, I've introduced the long requested `"inverse"` option which lets a user define the association on both sides in one call.

Instead of doing this:

```typescript
User.belongsToMany(Countries, { as: 'countries' });
User.belongsToMany(Countries, { as: 'citizen' });
```

users should now do this:

```typescript
User.belongsToMany(Countries, { as: 'countries', inverse: { as: 'citizen' } });
```

Closes https://github.com/sequelize/sequelize/issues/5144

## Includes are now strict

This PR normalizes Includes, which resolves a series of inconsistencies and should make them more stable.

For instance, `Includes` are always validated. If no association is found matching that include, an error will be thrown (including nested includes).

## ~PENDING QUESTION: Found an issue with `BelongsToMany`.~

Edit: resolved, see comments

Given a `BelongsToMany` association from `User` to `Project` through `User_Project`,
and using `groupedLimit` on the `BelongsToMany` association,

- [This unit test](https://github.com/sequelize/sequelize/blob/main/test/unit/sql/select.test.js#L110-L134) expects the through table to be aliased to `User_Project`.
- [This integration test](https://github.com/sequelize/sequelize/blob/main/test/integration/model/findAll/groupedLimit.test.js#L76-L93) expects the through table to be aliased to `Project`.

This needs to be unified, although I'm not sure which choice is best. Right now I've picked the second behavior.

## Other breaking changes

- The options 'foreignKeyConstraint' and 'constraints' have been merged in 'foreignKeyConstraints' as they did the same thing.
- The 'timestamps' option in `BelongsToManyOptions` has been moved to `ThroughOptions.timestamps`
- The 'uniqueKey' option in `BelongsToManyOptions` has been merged with `ThroughOptions.unique`
- HasMany / HasOne don't use 'as' as the foreign key anymore (it wouldn't make sense. You ended up with the class `Project.hasMany(User)` meant the model `User` received a foreign key called `UserId`.
- The `foreignKey` and `otherKey` options are unified with `ColumnOptions`:
   - The 'keyType' option is now 'foreignKey.type'
   - The 'onDelete' option is not 'foreignKey.onDelete'
   - The 'onUpdate' option is not 'foreignKey.onUpdate'
- The default value for `onDelete` for non-null FKs is unified. It's now always `CASCADE`. Previously, it was `NO ACTION` if defined first through `belongsTo`, and `CASCADE` if defined first through `hasOne` or `hasMany`, leading to unexpected behaviors depending on which method was called first.

Closes https://github.com/sequelize/sequelize/issues/13429
Closes https://github.com/sequelize/sequelize/issues/12782
Closes https://github.com/sequelize/sequelize/issues/11837

# What's left to be done

- [ ] Merge https://github.com/sequelize/sequelize/pull/14333
- [x] Merge https://github.com/sequelize/sequelize/pull/14359
- [x] Merge https://github.com/sequelize/sequelize/pull/14360
- [x] Add more typing tests
- [x] Check 'inverse' supports all options
- [x] add configuration for `fromThroughToSource` and `fromSourceToThrough`
- [x] Add support for `inverse` in `HasMany`, `HasOne`, `BelongsToMany`
- [x] Add support for `inverse` in `BelongsTo`
- [x] Determine the best option for the issue 'Found an issue with `BelongsToMany`.'

# What's next (in follow-up PRs)?

- [ ] Consider renaming `otherKey` to `inverse.foreignKey`
- [ ] Forbid defining an association if the primary key is composite. As this is not supported and breaks silently. (we can add support for this but users should be aware that it will not work).
- [ ] Document the breaking changes in the upgrade guide
- [ ] Update the documentation to reflect new recommended behaviors (`includes`, super many-to-many, `inverse`)
- [ ] Document the best practices to avoid errors with associations: Use 'inverse' instead of defining the association twice, beware of creating the same association twice without naming the inverse (`this.User.belongsToMany(this.Task) +  this.User.belongsToMany(this.Task, {as:'association2'})`)
- [ ] Improve [`ORDER BY`](https://github.com/sequelize/sequelize/tree/ephys/order-typing) thanks to the normalization of associations

### CODE REVIEW

File: test/unit/esm-named-exports.test.js
Comment (ephys) [reply test/unit/esm-named-exports.test.js:131]: I removed it by accident 😅

**Comment by ephys** [comment]
New breaking change to continue with working on making associations safe by default:

Calling `Project.belongsToMany(User)` and `User.belongsToMany(Project)` with different options in each call is not allowed anymore. It lead to unpredictable behavior.

This was done because `belongsToMany` is the only association that is automatically defined on both the source & target models.

Doing so results in the following error being thrown:

```
AssertionError [ERR_ASSERTION]: Project.belongsToMany(User) was called with different options than User.belongsToMany(Project). This is not allowed.
Note that 'belongsToMany' associations are automatically created on the target model as well, so you only need to call this method on one side.
```

**Comment by ephys** [comment]
Crazy that this is allowed, and even used in tests:

image

This defined two different associations with the same name. `User.associations.tasks` was equal to the most recent call to `belongsToMany`

After this PR, this just throws an error explaining that the association name is already in use for that model, and to use `as` to rename it. The updated code looks like this:

image

**Comment by WikiRik** [comment]
About `BelongsToMany`; is there any reason why you chose for the second? In general I would say that `User_Project` is clearer naming so that would have my preference but I haven't looked into it

**Comment by ephys** [comment]
Honestly the more I look at that, the less I know what the right answer is

The query generator transforms `groupedLimit` on belongsToMany associations into a `groupedLimit` on a `hasOne` association between the model and the belongsToMany's through table: https://github.com/sequelize/sequelize/blob/ephys/ts-associations/src/dialects/abstract/query-generator.js#L1282

We could:
- lie and pretend it's the belongsToMany association (name the join `project`)
- continue with the hasOne association (name it `userProjects`)
- use the same 'hack' other parts of the code use that explicitely rename that hasOne association to the name of the through model.

BelongsToMany are weird but mostly it's for good reasons. I think we should go with the third option after all and name it `User_Project` (the name of the through model)

**Comment by ephys** [comment]
My latest commit changes it back to `User_Project`, which makes it aligned with every other place the through model is automatically included

**Comment by sdepold** [comment]
Oh ... my ... god ... 🙈😳👌

**Comment by ephys** [comment]
If you're looking for pointers to review this huge PR, here is what I would recommend:

- Go through the list of changes described in the original post to see if you agree with them. It's not because something is implemented that it can't be reversed
- Go through the tests first: see if the changes still make sense to you
- Then the actual source code 😅

**Comment by ephys** [comment]
Let me know when you're ready to review, I'll resolve the conflicts then

**Comment by WikiRik** [comment]
I took the time a few days back to look at the changes and they look good to me. It will cause some time to upgrade projects so we need to be mindful with the examples we give in the upgrade guide (and in general how we document the new best practices in using associations in the various pages)

Next step for me would indeed be reviewing the tests but that will take some time. Would be nice if you could resolve the conflicts here once #14333 has been merged

**Comment by ephys** [comment]
> so we need to be mindful with the examples we give in the upgrade guide 

yup! Ideally I'd like to do that & rewrite the documentation on association before we merge, but I don't know when I'll be able to

> Would be nice if you could resolve the conflicts here once https://github.com/sequelize/sequelize/pull/14333 has been merged

will do

**Comment by ephys** [comment]
Conflicts resolved

File: src/associations/has-one.ts
Comment (WikiRik) [inline src/associations/has-one.ts:103]: When should we work on this, before the release of v7 or after?

File: src/associations/has-one.ts
Comment (WikiRik) [inline src/associations/has-one.ts:459]: Is this for before the release of v7 or after?

File: src/associations/has-many.ts
Comment (WikiRik) [inline src/associations/has-many.ts:199]:
```suggestion
    // TODO: scopes should be combined using AND instance of overwriting.
```
Is this for this PR, before the release of v7 or after?

File: src/associations/has-many.ts
Comment (WikiRik) [inline src/associations/has-many.ts:325]: Will this be addressed in #14020 ?

File: src/associations/has-many.ts
Comment (WikiRik) [inline src/associations/has-many.ts:505]: When do you think we should take care of this? Before or after the release of v7?

File: src/associations/belongs-to-many.ts
Comment (WikiRik) [inline src/associations/belongs-to-many.ts:131]: Where is that '!' added?

File: src/associations/belongs-to-many.ts
Comment (WikiRik) [inline src/associations/belongs-to-many.ts:446]: Just for completeness;  I made a comment about this in one of the other files as well. That same comment applies here

File: src/associations/belongs-to-many.ts
Comment (WikiRik) [inline src/associations/belongs-to-many.ts:548]: Just for completeness; I made a comment about this in one of the other files as well. That same comment applies here

File: src/associations/belongs-to-many.ts
Comment (WikiRik) [inline src/associations/belongs-to-many.ts:737]: Do we want to make this a TODO for a future PR if somebody wants to fix this properly?

File: src/associations/belongs-to-many.ts
Comment (WikiRik) [inline src/associations/belongs-to-many.ts:880]: When should we take care of this?

Comment (ephys) [reply src/associations/belongs-to-many.ts:880]: Oh, it's done. That's an old TODO. Removed

File: test/integration/hooks/associations.test.js
Comment (ephys) [reply test/integration/hooks/associations.test.js:93]: I didn't feel comfortable removing a test but this is one last good opportunity to check that the test was broken and that we can delete this

File: src/model.d.ts
Comment (ephys) [reply src/model.d.ts:2233]: I would suggest doing it during the migration of model.js to TypeScript. Doesn't really matter if we do it before or after the release of v7, we'll probably deprecate `hasAlias` and add `hasAssociation` to avoid another breaking change

File: src/associations/belongs-to.ts
Comment (ephys) [reply src/associations/belongs-to.ts:97]: It's not urgent. Could be for v7, could be for v8. We could also add support for composite foreign keys

File: src/associations/belongs-to.ts
Comment (ephys) [reply src/associations/belongs-to.ts:269]: It was already like this in v6. We should investigate whether it's intended or not but either way it's not a priority

File: src/associations/belongs-to-many.ts
Comment (ephys) [reply src/associations/belongs-to-many.ts:131]: Stale comment, I've removed it

File: src/associations/has-many.ts
Comment (ephys) [reply src/associations/has-many.ts:505]: Either is fine. It wouldn't be a breaking change if done properly

File: test/support.ts
Comment (ephys) [reply test/support.ts:348]: merged

**Review by WikiRik — APPROVED** [review]
I still have some questions open, but nothing that's blocking this PR. Good work!

**Review by WikiRik — CHANGES_REQUESTED** [review]
Woops, it seemed that merging main broke some unit tests. Should have looked at that as well, before approving 😅

**Comment by ephys** [comment]
> I mainly have questions to when TODOs will have to be taken care of

Would it be enough if I mentioned whether the change was a breaking change? As much as I'd like to have some of them fixed before v7, I can't guarantee it with the time I can allocate to this project :/

**Review by WikiRik — APPROVED** [review]

**Comment by bminer** [comment]
Imagine you have a `Task` belongs to `Project` association. What if a Task had a `similarProjectID` field? How would this association work in Sequelize v7?

**Comment by bminer** [comment]
@WikiRik @ephys - Any ideas on my prior question? If table `tasks` has a `projectID` and a `similarProjectID`, how can you create both associations in Sequelize v7?

**Comment by bminer** [comment]
Please ignore my prior comment. I misunderstood this change, sorry! I'm going to hide it.

**Pull request merged**
